File: PPC/Assignments/KnapsackProblem2/KnapsackProblemGo/knapsack/knapsackGA.go
package knapsack

import (
	"math/rand"
)

const (
	NGenerations   = 500
	PopSize        = 100000
	ProbMutation   = 0.5
	TournamentSize = 3
)

func Tournament(r *rand.Rand, population [PopSize]*Individual) *Individual {
	best := population[r.Intn(PopSize)]

	for i := 0; i < TournamentSize; i++ {
		other := population[r.Intn(PopSize)]
		if other.Fitness > best.Fitness {
			best = other
		}
	}
	return best
}

File: PPC/Assignments/KnapsackProblem2/KnapsackProblemRust/src/knapsack/knapsack_ga.rs
use rand::Rng;
use super::individual::Individual;

pub const N_GENERATIONS: usize = 500;
pub const POP_SIZE: usize = 100000;
pub const PROB_MUTATION: f64 = 0.5;
const TOURNAMENT_SIZE: usize = 3;

pub fn tournament<'a>(r: &mut impl Rng, population: &'a [Individual]) -> &'a Individual {
    let mut best = &population[r.random_range(0..population.len())];

    for _ in 1..TOURNAMENT_SIZE {
        let other = &population[r.random_range(0..population.len())];
        if other.fitness > best.fitness {
            best = other;
        }
    }

    best
}